kaitai_utils.py

```diff
@@ -37,16 +39,66 @@ def get_plugin_title_from_metadata():
         return _to_camel_case(metadata["id"])
 
 
-def write_to_json(data_binary: BinaryIO, writer: BufferedWriter, class_type: Type[KaitaiStruct]):
-    """
-    Writes a binary form of JSON string into a BufferedWriter
-
-    @param data_binary: binary data containing the file content
-    @param writer: bufferedWriter to write the binary form of the JSON string to
-    @param class_type: class that contains the parsing to a KaiTai struct
-    @return: JSON string representing contents of data object
-    """
-    writer.write(bytes(to_json_string(data_binary, class_type), "utf-8"))
+class JsonWriter:
+    def __init__(self, writer: BufferedWriter, trace: ExtractionTrace):
+        self.writer = writer
+        self.trace = trace
+
+    @streamable_list
+    def _list_to_dict(self, object_list: List[Any], path: str) -> Generator[
+        tuple[str, Any], None, None]:
+        for value_index, value in enumerate(object_list):
+            yield self._object_to_dict(value, path + f'.[{value_index}]')
+
+    @streamable_dict
+    def _object_to_dict(self, instance: Any, path: str) -> Generator[Dict[str, Any], None, None]:
+        """
+        Recursive helper method that parses an object to a dictionary.
+        Key: The parameters and property method names
+        Value: The parsed value or returning values of the fields and property method names
+
+        @param instance: object that needs parsing to dictionary
+        @param path: string representing the jsonpath to the current node in the object tree
+        @return: dictionary containing parsed fields and their respective parsed values in a dictionary
+        """
+        parameters_dict = _parameters_dict(instance)
+        for key, value_object in parameters_dict.items():
+            if is_public_property(key, value_object):
+                if _is_kaitai_struct(value_object):
+                    yield _to_lower_camel_case(key), self._object_to_dict(value_object, path + '.' + key)
+                elif _is_list(value_object):
+                    yield _to_lower_camel_case(key), self._list_to_dict(value_object, path)
+                elif isinstance(value_object, bytes):
+                    if len(value_object) < hansken_extraction_plugin.runtime.constants.MAX_CHUNK_SIZE:
+                        child_builder = self.trace.child_builder(path)
+                        child_builder.update(data={'raw': value_object}).build()
```

Comment:
**out of scope**, but i am not a big fan of this update method in a builder....it is a confusing read

a builder is meant for constructing an object by setting the fields of your choice of an object. In my opinion updating in a builder violates the intention of builder pattern.

Is it an idea to change this in another PR something similar as the following?

```
Trace child = child_builder.data({'raw': value_object}).build() 
self.trace.addChild(child)
```

it might be that the `child_builder` does not have a `data` method and `trace` does not have an `addChild` method, but you might get the idea.

Reply:
This is not related to the Kaitai plugin, but to the Extraction Plugin SDK itself.
The SDK follows the Hansken.py API. Hansken.py has a `TraceBuilder` with an `update` and `build` method. The `update` updates _the internal state_ of the builder, not a non-build resource.

Reply:
Discussed live and decided not to do it.